---
title: "**QDNAseq**"
author: "Michael Bell,  15002484" 
date: "05/01/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# **Quantitative DNA Sequencing for Chromosomal Aberrations**
## **QDNAseq R Package**

## **Contents:**  
**1:** Introduction  
\ **1.1:** Chromosomal Aberrations  
\ **1.2:**   
\ **1.3:**  
\ **1.4:**  
**2:** Running QDNAseq  
\ **2.1:** Bin Annotations  
\ **2.2:** Processing BAM Files  
\ **2.3:** Downstream Analyses  

## **1: Introduction:**
Scheinin et al. (2014) developed the QDNAseq pipeline to help improve the detection of DNA copy number aberrations from whole-genome sequencing (WGS). The focus of the study was to improve analysis and quantification of the challenges that are presented by WGS; this includes reference genome errors, sequence completions, repeat sequences, polymorphisms, variability in sample quality and procedure bias.  

## **1.1: Chromosomal Aberrations:**  
Chromosomal aberrations are defined as a change within either the structure or number of a chromosome. Most typical chromosomal aberrations are considered to be aneuploids; meaning they contain or are missing a number of chromosomes (e.g trisomy and monosomy) (Abhishek et al. (2018). 

## **1.2: **

## **1.3: **

## **1.4: **

## **2: Running QDNAseq**

## **2.1: Bin Annotations**
```{r bin annotations}
data(LGG150) # sample data for purpose of study 

readCounts <- (LGG150)

readCounts

plot(readCounts, logTransform=FALSE, ylim=c(-50, 200))

readCountsFiltered <- applyFilters(readCounts, residual=TRUE, blacklist=TRUE)

isobarPlot(readCountsFiltered)
```
## **2.2: ** 

## **2.3: **




```{r cars}
summary(cars)
```

## Including Plots
 
You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
